[Zeli0] iP

Zeli0 opened the pull request:
# Skylar

> "My name is Skyler White yo" - Skyler White ([source](https://knowyourmeme.com/memes/my-name-is-skyler-white-yo))

Thus speaks Skylar, an advanced **chat-bot** here to make your life easier. :bowtie:
Skylar is text-based, and helps to record and organize any tasks that you need to do, all while responding in an entertaining manner!

## Features

Skylar is a sophisticated chat-bot with various features!

+ Record various types of tasks
  + Things to do
  + Events with time ranges
  + Deadlines with an end data
+ Search for all the relevant tasks with a keyword
+ Easily delete entries from the list of tasks
+ Load and store the list

## Setting up Skylar

Downloading and using Skylar is easy! Just follow these steps:

1. Download the latest jar from [this link](https://github.com/Zeli0/ip/releases/tag/A-Jar).
2. Double click to run.
3. Begin adding tasks!

## For Programmers

If you are looking to learn from or build upon this project, you can take a look inside. The `main` method that starts everything up is here:

```
public class Hoster {
    public static void main(String[] args) {
        Application.launch(Main.class,args);
    }
```

## Future Development

More features will be implemented in the future! These include but are not limited to:

- [ ] Detailed, expandable task descriptions
- [ ] Editing features, like rescheduling
- [ ] View tasks in a schedule

Commit:
Add Gradle support

Zeli0 committed:
Implemented skeleton of Duke.java

The chatbot planned to be made only prints out a logo.

Let's implement the backbone of the chatbot that will be built upon later.

Zeli0 committed:
Updated user input display

Zeli0 committed:
Added list functionality to Duke.java

For the chatbot to perform its purpose as a to-do list, it must be able to record data.

Let's implement that list functionality before further development.

Zeli0 committed:
Added marking functionality to Duke.java

To create a functional to-do list, the chatbot needs to alter the status of entries in the list.

Let's implement a basic "completed" toggle for the chatbot before advancing to more complex functions.

Zeli0 committed:
Diversified todo entries

To increase the functionalities of the chatbot to-do list, extra information needs to be present in each entry.

Let's start by adding extra types of tasks for more nuance.

Zeli0 committed:
Added automated testing

To make sure the program works as intended, checks need to be made to the program.

Let's implement an automated testing system to simplify the checking process.

Zeli0 committed:
Marked potential errors for handling

Zeli0 committed:
Added error checking

In order for the todo list bot to function as desired, it needs to check for any potential errors.

Lets implement some of these errors to improve user experience.

Zeli0 committed:
Added delete function

The todo list bot needs to be able to remove tasks that the user does not want.
Let's implement this with a delete function.

Zeli0 committed:
Added saving function

To further improve the todo list bot, it needs to be able to store the list permanently.
Let's implement this saving function.

Zeli0 committed:
Added date detection

To further improve the sophistication of the to-do list chatbot, it should be able to automatically interpret certain inputs for the convenience of users.
Let's implement a preliminary version with date detection.

Zeli0 committed:
Merge branch 'branch-Level-7'

# Conflicts:
#	src/main/java/Duke.java

NoyaRoeT commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,23 @@
+public class Deadline extends Task {
+    String dead;
+    LocalDate deadDate;
+    public Deadline(String details, String dead) {
```
The parameter name dead is a bit confusing, considering it has the meaning of death. Perhaps a more intuitive name here?

NoyaRoeT commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,244 @@
+import java.util.ArrayList;
+import java.util.Locale;
+import java.util.Scanner;
+<<<<<<< HEAD
```
It would probably be better to remove line 4 and other github inserts.

NoyaRoeT commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,244 @@
+        }
+    }
+
+    private static boolean processInput(String input, ArrayList<Task> list, boolean end) {
```
It does not seem like there's any need for `end` to be passed in here. Perhaps just initializing and returning it at `case "bye":` is enough?

NoyaRoeT commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,244 @@
+        return end;
+    }
+
+    private static boolean byeOperation(String[] inputAnalyzed, ArrayList<Task> list){
```
In terms of code quality/readability, it may be better to use verbs for method names.

NoyaRoeT commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,244 @@
+        try {
+            String[] deadlineAnalyze = input.split("/by");
+            String deadline = deadlineAnalyze[1].trim();
+            String deets = deadlineAnalyze[0].split("deadline")[1].trim();
```
Perhaps a more intuitive name than `deets` would be better here?

NoyaRoeT reviewed (commented):
Overall, the code was structured logically and it was easy to follow the flow of the program. However, the variables names were a bit confusing which made the code harder to parse at a lower level.

Zeli0 committed:
Improving OOP

To make the todo list chatbot more modular and easier to understand, reformatting the project with more object-oriented principles may help.
Lets implement this by reforming the chatbot with new classes.

Zeli0 committed:
Organized files into packages

To further assist with understanding the different components of the chatbot, files can be divided into packages grouped based on their purpose.
Let's implement this in the chatbot project.

Zeli0 committed:
Merge remote-tracking branch 'upstream/add-gradle-support'

Zeli0 committed:
Added Gradle

To help make some tasks in the project easier, gradle can be added to automate some of the tasks.
Let's add gradle to the project.

Zeli0 committed:
Add test cases

To make sure the methods are working as designed, test cases need to be added.
Let's implement these with the newly added Gradle functionalities.

chunzkok commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,44 @@
+import command.Parser;
```
I like how you explicitly list the individual classes to be imported -- it makes it easier to see which exact classes are being used in this file :)

chunzkok commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,44 @@
+
+import java.io.IOException;
+
+import static java.lang.Integer.parseInt;
```
I'm not sure if this is actually needed as Java automatically imports the entire `java.lang.*` package -- maybe consider removing this and see if it still compiles?

chunzkok commented on src/main/java/command/InvalidInputException.java:
```diff
@@ -0,0 +1,7 @@
+package command;
+
+public class InvalidInputException extends Exception{
```
Might want to add a space between `...extends Exception` and the curly bracket -- just a small coding style thing

chunzkok commented on src/main/java/command/Parser.java:
```diff
@@ -0,0 +1,173 @@
+
+    public void processInput(String input) {
+        String[] inputAnalyzed = input.split(" ");
+        if (input.contains("#")) {
+            ui.showInvalidInputError("Illegal character '#'.");
+            return;
```
Maybe we can throw an `InvalidInputException` for this?

chunzkok commented on src/main/java/command/Parser.java:
```diff
@@ -0,0 +1,173 @@
+                default:
+                    ui.unknownCommand();
+            }
+        } catch (InvalidInputException e) {
+            ui.showInvalidInputError(e.getMessage());
+        } catch (IndexOutOfBoundsException e) {
+            ui.showArrayOutOfBoundsError();
+        } catch (NumberFormatException e) {
+            ui.showNumberFormatError();
```
Nice! Catching specific exceptions instead of one catching one general `Exception` will definitely help in debugging

chunzkok commented on src/main/java/command/Ui.java:
```diff
@@ -0,0 +1,119 @@
+
+    public void startMessage() {
+        dashedLine();
+        String intro = "My name is Skyler White yo \nHow can I help you?";
```
Hahaha nice name!

chunzkok commented on src/main/java/task/Task.java:
```diff
@@ -0,0 +1,32 @@
+package task;
+
+public class Task {
```
Maybe we can make this abstract? Since we won't really instantiate this class directly if i'm not wrong

chunzkok commented on src/main/java/task/TaskList.java:
```diff
@@ -0,0 +1,58 @@
+        }
+    }
+
+    public TaskList() {
```
Perhaps this can be removed since it's empty!

chunzkok commented on text-ui-test/input.txt:
```diff
@@ -0,0 +1,6 @@
+fuck
```
AYO WAD DIS 😨

chunzkok reviewed (commented):
I like how you gave your bot a personality! 
Your input validation is also very well thought-out, I will have to learn from how you considered the many different invalid inputs. 👍

Zeli0 committed:
Added Javadoc

To facilitate an easier coding experience, descriptors need to be added to each method so that others know what they are for.
Let's implement this with Javadoc.

Zeli0 committed:
Fix code to coding standards

Zeli0 committed:
Add find method

To make the chatbot even more robust, more functionalities need to be
added.

Let's start by implementing a "find" function.

Zeli0 committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/command/Parser.java

Zeli0 committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/command/Parser.java

Zeli0 committed:
Add GUI Base

To transition to a more sophisticated chatbot, our Duke needs to be more
user friendly.
Let's start by implementing the backbone of a new GUI.

Zeli0 committed:
Add GUI Backbone

To further implement a sophisticated GUI, more features need to be added.

Let's start by beginning the replacement of the existing UI with GUI.

Zeli0 committed:
Build on GUI

As the Javafx library is encountering numerous issues, steps need to be
taken to ensure that it works correctly.

Let's attempt to address them.

Zeli0 committed:
Update gitignore

Zeli0 committed:
Continue transition to JavaFX GUI

There still are some issues with the integration of JavaFX into the Duke
chatbot.
Let's fix these issues and integrate JavaFX formally into Duke.

Zeli0 committed:
Finish GUI implementation

To fully make the Duke chatbot sophisticated, a complete, functional GUI
needs to be added.
Let's implement this with JavaFX and FXML, ensuring that it is
completely integrated with the existing functionalities.

Zeli0 committed:
Add preliminary assertions

To ensure the chatbot works as intended, checks need to be added to make
sure invariants are not broken.

As a start to the checking process, let's implement some preliminary
assertion checks at key points of the program.

Zeli0 committed:
Add personality

To further cement the uniqueness of this chatbot, personality is needed
in its messages and name.

Let's finalize the first version of the personality implementation,
together with all past personality additions, by changing the name
of the bot formally to "Skylar".

Zeli0 committed:
Improve code quality

To further polish up the Skylar chatbot, the readibility of the code
is important for further iterations.

Let's improve readibility by improving code quality.

Zeli0 committed:
Improve code quality of TaskList.java

Zeli0 committed:
Add more assertions

To check to ensure that invariants are preserved, more assertions need
to be made.

Let's implement these assertions in the event classes.

Zeli0 committed:
Merge pull request #2 from Zeli0/branch-A-CodeQuality

Merge code quality branch

Zeli0 committed:
Merge branch 'master' into branch-A-Assertions

Zeli0 committed:
Merge pull request #3 from Zeli0/branch-A-Assertions

Merge in assertion branch

Zeli0 committed:
Improve GUI

To improve user experience, building new features on the existing GUI
is necessary.

Let's start by labeling error messages so users know when they have an
incorrect input.

Zeli0 committed:
Begin help command implementation

Zeli0 committed:
Add extensive features

To further improve the quality of the Skylar chatbot, advanced functionalities need to be implemented.

Let's start by implementing a new help function, and an undo function.

Zeli0 committed:
Add User Guide

To help the user understand how to use the Skylar chatbot, a user guide is a necessary component of our project.

Let's write one with all the current features and how they should be used.

Zeli0 committed:
Give main window title

Zeli0 committed:
Fix undo

Zeli0 committed:
Fix UI

Zeli0 committed:
Add product screenshot

Zeli0 committed:
Fix storage and undo clashes

Zeli0 committed:
Fix program running issues
